addons/sale_management/static/src/fields/sale_order_line_field/sale_order_line_field.js

```diff
@@ -10,12 +10,38 @@ patch(SaleOrderLineListRenderer.prototype, {
         this.copyFields.push('is_optional');
     },
 
-    get showOptionalButton() {
-        if (this.isSubSection(this.record)) {
-            const parentRecord = getParentSectionRecord(this.props.list, this.record);
-            return !parentRecord?.data?.is_optional;
-        }
-        return true;
+    /**
+     * Disable "Hide Composition" and "Hide Prices" buttons for optional sections and their
+     * subsections.
+     */
+    get disableCompositionButton() {
+        return (
+            super.disableCompositionButton
+            || this.record.data.is_optional
+            || this.shouldCollapse(this.record, 'is_optional')
+        );
+    },
+
+    get disablePricesButton() {
+        return (
+            super.disablePricesButton
+            || this.record.data.is_optional
+            || this.shouldCollapse(this.record, 'is_optional')
+        );
+    },
+
+    /**
+     * Disable "Set Optional" button if
+     *  - Parent section is optional
+     *  - Parent section hides prices or composition
+     *  - Section itself hides prices or composition
+     */
+    get disableOptionalButton() {
+        return (
+            this.shouldCollapse(this.record, 'is_optional')
+            || this.shouldCollapse(this.record, 'collapse_prices', true)
```

Comment:
to improve readability
```suggestion
            || this.shouldCollapse(this.record, 'collapse_prices', checkTopSection=true)
```